File: examples_eem.py
"""
Ejemplos de uso del módulo fluorescence_eem

Este archivo demuestra cómo usar las funciones para aplanar y procesar
matrices de espectros de fluorescencia de excitación-emisión.
"""

import numpy as np
from fluorescence_eem import (
    FluorescenceEEM,
    flatten_eem_matrix,
    reconstruct_eem_matrix,
    flatten_eem_batch
)


def ejemplo_basico():
    """Ejemplo básico de aplanamiento de matriz EEM"""
    print("=" * 60)
    print("EJEMPLO 1: Aplanamiento básico de matriz EEM")
    print("=" * 60)

    # Crear una matriz EEM de ejemplo (10 excitaciones x 15 emisiones)
    n_ex = 10
    n_em = 15
    data = np.random.rand(n_ex, n_em) * 100

    print(f"\nMatriz original: {data.shape}")
    print(f"Primeros valores:\n{data[:3, :5]}\n")

    # Aplanar la matriz
    flat_data = flatten_eem_matrix(data, order='C')
    print(f"Vector aplanado: {flat_data.shape}")
    print(f"Primeros 10 valores: {flat_data[:10]}\n")

    # Reconstruir la matriz
    reconstructed = reconstruct_eem_matrix(flat_data, n_ex, n_em, order='C')
    print(f"Matriz reconstruida: {reconstructed.shape}")
    print(f"Primeros valores:\n{reconstructed[:3, :5]}\n")

    # Verificar que la reconstrucción es correcta
    is_equal = np.allclose(data, reconstructed)
    print(f"¿Reconstrucción exitosa? {is_equal}\n")


def ejemplo_con_longitudes_de_onda():
    """Ejemplo usando longitudes de onda reales"""
    print("=" * 60)
    print("EJEMPLO 2: EEM con longitudes de onda reales")
    print("=" * 60)

    # Longitudes de onda típicas en espectroscopía de fluorescencia
    excitation_wl = np.arange(250, 450, 5)  # 250-445 nm, cada 5 nm
    emission_wl = np.arange(300, 600, 5)    # 300-595 nm, cada 5 nm

    # Simular datos de fluorescencia con un pico
    n_ex = len(excitation_wl)
    n_em = len(emission_wl)
    data = np.zeros((n_ex, n_em))

    # Crear un pico gaussiano en ex=350nm, em=450nm
    for i, ex in enumerate(excitation_wl):
        for j, em in enumerate(emission_wl):
            # Pico gaussiano
            intensity = 100 * np.exp(-((ex - 350)**2 / (2 * 20**2) + (em - 450)**2 / (2 * 30**2)))
            data[i, j] = intensity

    # Crear objeto EEM
    eem = FluorescenceEEM(data, excitation_wl, emission_wl)
    print(f"\n{eem}")

    # Obtener estadísticas
    stats = eem.get_stats()
    print(f"\nEstadísticas:")
    for key, value in stats.items():
        print(f"  {key}: {value}")

    # Aplanar con etiquetas
    flat_data, labels = eem.flatten_with_labels()
    print(f"\nDatos aplanados: {len(flat_data)} puntos")
    print(f"Primeras 5 etiquetas (Ex, Em): {labels[:5]}")
    print(f"Primeros 5 valores de intensidad: {flat_data[:5]}\n")


def ejemplo_normalizacion():
    """Ejemplo de normalización de EEM"""
    print("=" * 60)
    print("EJEMPLO 3: Normalización de matrices EEM")
    print("=" * 60)

    # Crear datos de ejemplo
    excitation_wl = np.arange(250, 450, 10)
    emission_wl = np.arange(300, 600, 10)
    n_ex, n_em = len(excitation_wl), len(emission_wl)

    # Datos con valores entre 50 y 150
    data = np.random.rand(n_ex, n_em) * 100 + 50

    eem = FluorescenceEEM(data, excitation_wl, emission_wl)
    print(f"\nDatos originales - min: {eem.get_stats()['min']:.2f}, max: {eem.get_stats()['max']:.2f}")

    # Normalización por máximo
    eem_norm_max = eem.normalize(method='max')
    print(f"Normalizado por max - min: {eem_norm_max.get_stats()['min']:.2f}, max: {eem_norm_max.get_stats()['max']:.2f}")

    # Normalización por área
    eem_norm_area = eem.normalize(method='area')
    print(f"Normalizado por área - suma total: {np.sum(eem_norm_area.data):.4f}")

    # Normalización z-score
    eem_norm_z = eem.normalize(method='zscore')
    print(f"Normalizado z-score - media: {eem_norm_z.get_stats()['mean']:.4f}, std: {eem_norm_z.get_stats()['std']:.4f}\n")


def ejemplo_eliminar_rayleigh():
    """Ejemplo de eliminación de dispersión de Rayleigh"""
    print("=" * 60)
    print("EJEMPLO 4: Eliminación de dispersión de Rayleigh")
    print("=" * 60)

    # Crear datos de ejemplo
    excitation_wl = np.arange(250, 500, 5)
    emission_wl = np.arange(250, 600, 5)
    n_ex, n_em = len(excitation_wl), len(emission_wl)

    data = np.random.rand(n_ex, n_em) * 50

    # Añadir dispersión de Rayleigh (alta intensidad cuando ex ≈ em)
    for i, ex in enumerate(excitation_wl):
        for j, em in enumerate(emission_wl):
            if abs(em - ex) < 15:
                data[i, j] = 1000  # Alta intensidad de dispersión

    eem = FluorescenceEEM(data, excitation_wl, emission_wl)
    print(f"\nDatos originales:")
    print(f"  Max: {eem.get_stats()['max']:.2f}")
    print(f"  Puntos con datos válidos: {np.sum(~np.isnan(eem.data))}")

    # Eliminar dispersión de Rayleigh
    eem_cleaned = eem.remove_rayleigh_scatter(bandwidth=15)
    print(f"\nDatos después de eliminar Rayleigh:")
    print(f"  Max: {np.nanmax(eem_cleaned.data):.2f}")
    print(f"  Puntos con datos válidos: {np.sum(~np.isnan(eem_cleaned.data))}")
    print(f"  Puntos eliminados (NaN): {np.sum(np.isnan(eem_cleaned.data))}\n")


def ejemplo_batch_processing():
    """Ejemplo de procesamiento por lotes de múltiples EEM"""
    print("=" * 60)
    print("EJEMPLO 5: Procesamiento por lotes de múltiples EEM")
    print("=" * 60)

    # Simular 5 muestras diferentes
    n_samples = 5
    n_ex, n_em = 20, 30

    eem_list = []
    for i in range(n_samples):
        data = np.random.rand(n_ex, n_em) * (i + 1) * 20  # Diferentes intensidades
        eem = FluorescenceEEM(data)
        eem_list.append(eem)

    print(f"\nNúmero de muestras: {n_samples}")
    print(f"Dimensiones de cada EEM: {n_ex} x {n_em}")

    # Aplanar todas las EEM en una matriz
    batch_matrix = flatten_eem_batch(eem_list, order='C')
    print(f"\nMatriz de lote: {batch_matrix.shape}")
    print(f"  (cada fila es una EEM aplanada)")

    # Mostrar estadísticas de cada muestra
    print(f"\nEstadísticas por muestra:")
    for i, eem in enumerate(eem_list):
        stats = eem.get_stats()
        print(f"  Muestra {i+1}: mean={stats['mean']:.2f}, max={stats['max']:.2f}")

    print()


def ejemplo_pandas_export():
    """Ejemplo de exportación a pandas DataFrame"""
    print("=" * 60)
    print("EJEMPLO 6: Exportación a pandas DataFrame")
    print("=" * 60)

    try:
        import pandas as pd

        # Crear datos de ejemplo pequeños
        excitation_wl = np.array([250, 300, 350])
        emission_wl = np.array([300, 350, 400, 450])
        data = np.random.rand(3, 4) * 100

        eem = FluorescenceEEM(data, excitation_wl, emission_wl)

        # Convertir a DataFrame
        df = eem.to_dataframe()
        print(f"\nDataFrame generado ({len(df)} filas):")
        print(df.head(10))
        print(f"\nColumnas: {list(df.columns)}")

        # Ejemplo de filtrado
        high_intensity = df[df['intensity'] > 50]
        print(f"\nPuntos con intensidad > 50: {len(high_intensity)} puntos")

    except ImportError:
        print("\n[NOTA] pandas no está instalado. Instalar con: pip install pandas")

    print()


def ejemplo_comparacion_ordenes():
    """Ejemplo comparando órdenes de aplanamiento 'C' vs 'F'"""
    print("=" * 60)
    print("EJEMPLO 7: Comparación de órdenes de aplanamiento")
    print("=" * 60)

    # Matriz pequeña para visualizar la diferencia
    data = np.array([
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
        [10, 11, 12]
    ])

    print(f"\nMatriz original (4 x 3):")
    print(data)

    eem = FluorescenceEEM(data)

    # Orden 'C' (por filas)
    flat_c = eem.flatten(order='C')
    print(f"\nAplanado orden 'C' (row-major):")
    print(flat_c)
    print("Recorre: [1,2,3,4,5,6,7,8,9,10,11,12]")

    # Orden 'F' (por columnas)
    flat_f = eem.flatten(order='F')
    print(f"\nAplanado orden 'F' (column-major):")
    print(flat_f)
    print("Recorre: [1,4,7,10,2,5,8,11,3,6,9,12]")

    print()


if __name__ == "__main__":
    print("\n" + "="*60)
    print("EJEMPLOS DE USO: fluorescence_eem")
    print("Procesamiento de matrices de espectros de fluorescencia")
    print("="*60 + "\n")

    ejemplo_basico()
    ejemplo_con_longitudes_de_onda()
    ejemplo_normalizacion()
    ejemplo_eliminar_rayleigh()
    ejemplo_batch_processing()
    ejemplo_pandas_export()
    ejemplo_comparacion_ordenes()

    print("="*60)
    print("EJEMPLOS COMPLETADOS")
    print("="*60)

File: examples_eem.R
# ============================================================================
# Ejemplos de uso del módulo fluorescence_eem.R
#
# Este archivo demuestra cómo usar las funciones para aplanar y procesar
# matrices de espectros de fluorescencia de excitación-emisión en R.
#
# Author: @menapehi
# ============================================================================

# Cargar el módulo
source("fluorescence_eem.R")

cat("\n")
cat("================================================================\n")
cat("EJEMPLOS DE USO: fluorescence_eem (R)\n")
cat("Procesamiento de matrices de espectros de fluorescencia\n")
cat("================================================================\n\n")


# ============================================================================
# EJEMPLO 1: Aplanamiento básico de matriz EEM
# ============================================================================

ejemplo_basico <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 1: Aplanamiento básico de matriz EEM\n")
  cat("================================================================\n\n")

  # Crear una matriz EEM de ejemplo (10 excitaciones x 15 emisiones)
  set.seed(42)
  n_ex <- 10
  n_em <- 15
  data <- matrix(runif(n_ex * n_em) * 100, nrow = n_ex, ncol = n_em)

  cat(sprintf("Matriz original: %d x %d\n", n_ex, n_em))
  cat("Primeros valores:\n")
  print(data[1:3, 1:5])
  cat("\n")

  # Aplanar la matriz
  flat_data <- flatten_matrix(data, byrow = TRUE)
  cat(sprintf("Vector aplanado: %d elementos\n", length(flat_data)))
  cat("Primeros 10 valores:\n")
  print(flat_data[1:10])
  cat("\n")

  # Reconstruir la matriz
  reconstructed <- reconstruct_matrix(flat_data, n_ex, n_em, byrow = TRUE)
  cat(sprintf("Matriz reconstruida: %d x %d\n", nrow(reconstructed), ncol(reconstructed)))
  cat("Primeros valores:\n")
  print(reconstructed[1:3, 1:5])
  cat("\n")

  # Verificar que la reconstrucción es correcta
  is_equal <- all.equal(data, reconstructed)
  cat(sprintf("¿Reconstrucción exitosa? %s\n\n", isTRUE(is_equal)))
}


# ============================================================================
# EJEMPLO 2: EEM con longitudes de onda reales
# ============================================================================

ejemplo_con_longitudes_de_onda <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 2: EEM con longitudes de onda reales\n")
  cat("================================================================\n\n")

  # Longitudes de onda típicas en espectroscopía de fluorescencia
  excitation_wl <- seq(250, 445, by = 5)  # 250-445 nm, cada 5 nm
  emission_wl <- seq(300, 595, by = 5)    # 300-595 nm, cada 5 nm

  # Simular datos de fluorescencia con un pico gaussiano
  n_ex <- length(excitation_wl)
  n_em <- length(emission_wl)
  data <- matrix(0, nrow = n_ex, ncol = n_em)

  # Crear un pico gaussiano en ex=350nm, em=450nm
  for (i in seq_along(excitation_wl)) {
    for (j in seq_along(emission_wl)) {
      ex <- excitation_wl[i]
      em <- emission_wl[j]
      # Pico gaussiano
      intensity <- 100 * exp(-((ex - 350)^2 / (2 * 20^2) + (em - 450)^2 / (2 * 30^2)))
      data[i, j] <- intensity
    }
  }

  # Crear objeto EEM
  eem <- FluorescenceEEM(data, excitation_wl, emission_wl)
  print(eem)
  cat("\n")

  # Obtener estadísticas
  stats <- get_eem_stats(eem)
  cat("Estadísticas:\n")
  cat(sprintf("  Min: %.4f\n", stats$min))
  cat(sprintf("  Max: %.4f\n", stats$max))
  cat(sprintf("  Media: %.4f\n", stats$mean))
  cat(sprintf("  Desv. estándar: %.4f\n", stats$sd))
  cat(sprintf("  Mediana: %.4f\n", stats$median))
  cat("\n")

  # Aplanar con etiquetas
  result <- flatten_eem_with_labels(eem)
  cat(sprintf("Datos aplanados: %d puntos\n", length(result$data)))
  cat("Primeras 5 etiquetas (Ex, Em):\n")
  print(head(result$labels, 5))
  cat("\nPrimeros 5 valores de intensidad:\n")
  print(result$data[1:5])
  cat("\n")
}


# ============================================================================
# EJEMPLO 3: Normalización de matrices EEM
# ============================================================================

ejemplo_normalizacion <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 3: Normalización de matrices EEM\n")
  cat("================================================================\n\n")

  # Crear datos de ejemplo
  set.seed(123)
  excitation_wl <- seq(250, 440, by = 10)
  emission_wl <- seq(300, 590, by = 10)
  n_ex <- length(excitation_wl)
  n_em <- length(emission_wl)

  # Datos con valores entre 50 y 150
  data <- matrix(runif(n_ex * n_em) * 100 + 50, nrow = n_ex, ncol = n_em)

  eem <- FluorescenceEEM(data, excitation_wl, emission_wl)
  stats_orig <- get_eem_stats(eem)
  cat(sprintf("Datos originales - min: %.2f, max: %.2f\n",
             stats_orig$min, stats_orig$max))

  # Normalización por máximo
  eem_norm_max <- normalize_eem(eem, method = "max")
  stats_max <- get_eem_stats(eem_norm_max)
  cat(sprintf("Normalizado por max - min: %.2f, max: %.2f\n",
             stats_max$min, stats_max$max))

  # Normalización por área
  eem_norm_area <- normalize_eem(eem, method = "area")
  cat(sprintf("Normalizado por área - suma total: %.6f\n",
             sum(eem_norm_area$data)))

  # Normalización z-score
  eem_norm_z <- normalize_eem(eem, method = "zscore")
  stats_z <- get_eem_stats(eem_norm_z)
  cat(sprintf("Normalizado z-score - media: %.6f, sd: %.6f\n\n",
             stats_z$mean, stats_z$sd))
}


# ============================================================================
# EJEMPLO 4: Eliminación de dispersión de Rayleigh
# ============================================================================

ejemplo_eliminar_rayleigh <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 4: Eliminación de dispersión de Rayleigh\n")
  cat("================================================================\n\n")

  # Crear datos de ejemplo
  set.seed(456)
  excitation_wl <- seq(250, 495, by = 5)
  emission_wl <- seq(250, 595, by = 5)
  n_ex <- length(excitation_wl)
  n_em <- length(emission_wl)

  data <- matrix(runif(n_ex * n_em) * 50, nrow = n_ex, ncol = n_em)

  # Añadir dispersión de Rayleigh (alta intensidad cuando ex ≈ em)
  for (i in seq_along(excitation_wl)) {
    for (j in seq_along(emission_wl)) {
      ex <- excitation_wl[i]
      em <- emission_wl[j]
      if (abs(em - ex) < 15) {
        data[i, j] <- 1000  # Alta intensidad de dispersión
      }
    }
  }

  eem <- FluorescenceEEM(data, excitation_wl, emission_wl)
  cat("Datos originales:\n")
  cat(sprintf("  Max: %.2f\n", max(eem$data, na.rm = TRUE)))
  cat(sprintf("  Puntos con datos válidos: %d\n", sum(!is.na(eem$data))))

  # Eliminar dispersión de Rayleigh
  eem_cleaned <- remove_rayleigh_scatter(eem, bandwidth = 15)
  cat("\nDatos después de eliminar Rayleigh:\n")
  cat(sprintf("  Max: %.2f\n", max(eem_cleaned$data, na.rm = TRUE)))
  cat(sprintf("  Puntos con datos válidos: %d\n", sum(!is.na(eem_cleaned$data))))
  cat(sprintf("  Puntos eliminados (NA): %d\n\n", sum(is.na(eem_cleaned$data))))
}


# ============================================================================
# EJEMPLO 5: Procesamiento por lotes de múltiples EEM
# ============================================================================

ejemplo_batch_processing <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 5: Procesamiento por lotes de múltiples EEM\n")
  cat("================================================================\n\n")

  # Simular 5 muestras diferentes
  set.seed(789)
  n_samples <- 5
  n_ex <- 20
  n_em <- 30

  eem_list <- list()
  for (i in seq_len(n_samples)) {
    data <- matrix(runif(n_ex * n_em) * i * 20, nrow = n_ex, ncol = n_em)
    eem_list[[i]] <- FluorescenceEEM(data)
  }

  cat(sprintf("Número de muestras: %d\n", n_samples))
  cat(sprintf("Dimensiones de cada EEM: %d x %d\n", n_ex, n_em))

  # Aplanar todas las EEM en una matriz
  batch_matrix <- flatten_eem_batch(eem_list, byrow = TRUE)
  cat(sprintf("\nMatriz de lote: %d x %d\n", nrow(batch_matrix), ncol(batch_matrix)))
  cat("  (cada fila es una EEM aplanada)\n")

  # Mostrar estadísticas de cada muestra
  cat("\nEstadísticas por muestra:\n")
  for (i in seq_along(eem_list)) {
    stats <- get_eem_stats(eem_list[[i]])
    cat(sprintf("  Muestra %d: media=%.2f, max=%.2f\n", i, stats$mean, stats$max))
  }
  cat("\n")
}


# ============================================================================
# EJEMPLO 6: Convertir a data.frame
# ============================================================================

ejemplo_dataframe_export <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 6: Exportación a data.frame\n")
  cat("================================================================\n\n")

  # Crear datos de ejemplo pequeños
  set.seed(101)
  excitation_wl <- c(250, 300, 350)
  emission_wl <- c(300, 350, 400, 450)
  data <- matrix(runif(3 * 4) * 100, nrow = 3, ncol = 4)

  eem <- FluorescenceEEM(data, excitation_wl, emission_wl)

  # Convertir a data.frame
  df <- eem_to_dataframe(eem)
  cat(sprintf("Data.frame generado (%d filas):\n", nrow(df)))
  print(head(df, 10))
  cat(sprintf("\nColumnas: %s\n", paste(names(df), collapse = ", ")))

  # Ejemplo de filtrado
  high_intensity <- df[df$intensity > 50, ]
  cat(sprintf("\nPuntos con intensidad > 50: %d puntos\n\n", nrow(high_intensity)))
}


# ============================================================================
# EJEMPLO 7: Comparación de órdenes de aplanamiento
# ============================================================================

ejemplo_comparacion_ordenes <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 7: Comparación de órdenes de aplanamiento\n")
  cat("================================================================\n\n")

  # Matriz pequeña para visualizar la diferencia
  data <- matrix(c(
    1, 2, 3,
    4, 5, 6,
    7, 8, 9,
    10, 11, 12
  ), nrow = 4, ncol = 3, byrow = TRUE)

  cat("Matriz original (4 x 3):\n")
  print(data)
  cat("\n")

  eem <- FluorescenceEEM(data)

  # Orden byrow = TRUE (por filas)
  flat_byrow <- flatten_eem(eem, byrow = TRUE)
  cat("Aplanado byrow = TRUE (por filas):\n")
  print(flat_byrow)
  cat("Recorre: [1,2,3,4,5,6,7,8,9,10,11,12]\n\n")

  # Orden byrow = FALSE (por columnas)
  flat_bycol <- flatten_eem(eem, byrow = FALSE)
  cat("Aplanado byrow = FALSE (por columnas):\n")
  print(flat_bycol)
  cat("Recorre: [1,4,7,10,2,5,8,11,3,6,9,12]\n\n")
}


# ============================================================================
# EJEMPLO 8: Uso de métodos print y summary
# ============================================================================

ejemplo_metodos_s3 <- function() {
  cat("================================================================\n")
  cat("EJEMPLO 8: Uso de métodos print y summary\n")
  cat("================================================================\n\n")

  # Crear objeto EEM
  set.seed(202)
  excitation_wl <- seq(250, 400, by = 10)
  emission_wl <- seq(300, 500, by = 10)
  n_ex <- length(excitation_wl)
  n_em <- length(emission_wl)
  data <- matrix(runif(n_ex * n_em) * 200, nrow = n_ex, ncol = n_em)

  eem <- FluorescenceEEM(data, excitation_wl, emission_wl)

  # Usar método print
  cat("Método print():\n")
  print(eem)
  cat("\n")

  # Usar método summary
  cat("Método summary():\n")
  summary(eem)
  cat("\n")
}


# ============================================================================
# EJECUTAR TODOS LOS EJEMPLOS
# ============================================================================

if (interactive() || !interactive()) {
  ejemplo_basico()
  ejemplo_con_longitudes_de_onda()
  ejemplo_normalizacion()
  ejemplo_eliminar_rayleigh()
  ejemplo_batch_processing()
  ejemplo_dataframe_export()
  ejemplo_comparacion_ordenes()
  ejemplo_metodos_s3()

  cat("================================================================\n")
  cat("EJEMPLOS COMPLETADOS\n")
  cat("================================================================\n")
}
